File: (04-06-2024)Find All Anagrams in a String/joshlin/solution.py
from collections import Counter
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        sSize, pSize = len(s), len(p)
        if pSize > sSize: return []
        sCount, pCount = Counter(s[:pSize]), Counter(p)
        left, right = 0, pSize - 1
        res = []
        while right < sSize:
            if sCount == pCount:
                res.append(left)
            if sCount[s[left]] > 1:
                sCount[s[left]] -= 1
            else:
                del sCount[s[left]]
            left += 1
            right += 1
            if right < sSize:
                sCount[s[right]] = sCount.get(s[right], 0) + 1
        return res

File: (04-06-2024)Find All Anagrams in a String/joshlin/solution.cpp
#include <unordered_map>
#include <vector>
#include <string>

class Solution {
public:
    std::vector<int> findAnagrams(std::string s, std::string p) {
        int sSize = s.length(), pSize = p.length();
        std::unordered_map<char, int> sCount, pCount;
        for (int i = 0; i < pSize; i++) {
            sCount[s[i]]++;
            pCount[p[i]]++;
        }
        std::vector<int> res = sCount == pCount ? std::vector<int>{0} : std::vector<int>{};
        int left = 0;
        for (int right = pSize; right < sSize; right++) {
            sCount[s[left]]--;
            if (sCount[s[left]] == 0) sCount.erase(s[left]);
            sCount[s[right]]++;
            left++;
            if (sCount == pCount) res.push_back(left);
        }
        return res;
    }
};